eclipse-che/che-operator
# feat: configure workspace security context for container builds

### What does this PR do?
Configures the `workspace.containerSecurityContext` field of the Che-Operator-owned DWOC depending on the value of `devEnvironments.disableContainerBuildCapabilities` in the Che Cluster CR.

If `devEnvironments.disableContainerBuildCapabilities` is set to `false`, then the DWOC's `workspace.containerSecurityContext` is configured as follows:
```YAML
containerSecurityContext:
  allowPrivilegeEscalation: false
  capabilities:
    add:
    - SETGID
    - SETUID
```

If `devEnvironments.disableContainerBuildCapabilities` is set to `true`, then the DWOC's `workspace.containerSecurityContext` field is cleared.

### Screenshot/screencast of this PR
n/a

### What issues does this PR fix or reference?
eclipse/che#21770


### How to test this PR?
Some extra automated tests were added for this PR.

In terms of manual testing, there are 2 relevant cases (the first one concerns the core functionality this PR adds, and the second one ensures that the appropriate SCC is used when testing Che on OpenShift):

#### Ensuring the Che-Operator-owned DWOC is correctly configured (Minikube)
1. Deploy an image of DevWorkspace-Operator that has [the changes from this PR](https://github.com/devfile/devworkspace-operator/pull/953) to your cluster. I have an image hosted at: `quay.io/aobuchow/dwo-container-security-context`. 
```
# In the DevWorkspace-Operator repo:
export DWO_IMG=quay.io/aobuchow/dwo-container-security-context

minikube addons enable ingress && make install_cert_manager && make install
```
2. [Build an image of Che-Operator](https://github.com/eclipse-che/che-operator#build-and-push-custom-che-operator-image) using the changes from this PR. There is already one hosted at `quay.io/aobuchow/che-operator:next`
3. Deploy Che-Operator using the image from step **1**: `chectl server:deploy --installer operator -p minikube --che-operator-image=quay.io/aobuchow/che-operator:next`
4. Ensure the Che-Operator-owned DWOC does not have the `workspace.containerSecurityContext` field by default. 
`oc describe dwoc -n eclipse-che`

```YAML
apiVersion: controller.devfile.io/v1alpha1  
config:  
 workspace:  
kind: DevWorkspaceOperatorConfig  
metadata:  
 creationTimestamp: "2022-10-26T20:09:23Z"  
 generation: 2  
 name: devworkspace-config  
 namespace: eclipse-che  
(...)
```


5. Set `devEnvironments.disableContainerBuildCapabilities` to `true` the Che Cluster CR:
6. Ensure the Che-Operator-owned DWOC now has the `workspace.containerSecurityContext` field configured with the correct container security context:
`oc describe dwoc -n eclipse-che`

```diff
 apiVersion: controller.devfile.io/v1alpha1  
 config:  
  workspace:  
+    containerSecurityContext:  
+      allowPrivilegeEscalation: false  
+      capabilities:  
+        add:  
+        - SETGID  
+        - SETUID  
 kind: DevWorkspaceOperatorConfig  
 metadata:  
  creationTimestamp: "2022-10-26T20:09:23Z"  
  generation: 2  
  name: devworkspace-config  
  namespace: eclipse-che  
(...)
```


7. Set `devEnvironments.disableContainerBuildCapabilities` to `false` the Che Cluster CR.
8. Ensure that the `workspace.containerSecurityContext` field has been cleared from the Che-Operator-owned DWOC:
`oc describe dwoc -n eclipse-che`

```YAML
apiVersion: controller.devfile.io/v1alpha1  
config:  
 workspace:  
kind: DevWorkspaceOperatorConfig  
metadata:  
 creationTimestamp: "2022-10-26T20:09:23Z"  
 generation: 2  
 name: devworkspace-config  
 namespace: eclipse-che  
(...)
```

#### Ensuring containers can be built on OpenShift with an SCC present that has higher priority than container-build
1. Install a version of DWO (with the [changes from this PR](https://github.com/devfile/devworkspace-operator/pull/953))  using OLM. The easiest way to do this is to `oc apply -f `apply the following catalogsource and then install DWO from OperatorHub:
```YAML
apiVersion: operators.coreos.com/v1alpha1
kind: CatalogSource
metadata:
  name: devworkspace-operator-catalog
  namespace: openshift-marketplace
spec:
  sourceType: grpc
  image: quay.io/aobuchow/devworkspace-operator-index:next
  publisher: Red Hat
  displayName: DevWorkspace Operator Catalog
  updateStrategy:
    registryPoll:
      interval: 5m
``` 
2. Deploy Che using the operator image built from the current PR:
`chectl server:deploy -p openshift --che-operator-image=quay.io/aobuchow/che-operator:next`
3. Set `devEnvironments.disableContainerBuildCapabilities` to `false` the Che Cluster CR. 
4. Follow the SCC related instructions [from this PR](https://github.com/devfile/devworkspace-operator/pull/679) though do not create the workspace that uses the `controller.devfile.io/scc` attribute yet. **Note:** For Step 3. _Let the user creating the DevWorkspace use the container-build SCC_ requires you to have a user other than `kube:admin` on your cluster. I personally create other users with [HTPasswd](https://access.redhat.com/documentation/en-us/openshift_container_platform/4.7/html-single/authentication_and_authorization/index#identity-provider-creating-htpasswd-secret_configuring-htpasswd-identity-provider)
5. Apply a [higher priority SCC](https://gist.github.com/l0rd/87337c08ecf414499681261f9fbcebe2) to the cluster 
6. As a user that has access to the container-build SCC (other than `kube:admin`), create a workspace that uses the `controller.devfile.io/scc` attribute (Step 4 from the [SCC PR](https://github.com/devfile/devworkspace-operator/pull/679))
7. Check the workspace deployment and ensure the `openshift.io/scc: container-build` annotation is set. 


### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)


### Reviewers

Reviewers, please comment how you tested the PR when approving it.
